hoomd/md/force.py

```diff
@@ -117,212 +118,145 @@ def additional_virial(self):
-        tvec (tuple): torque vector :math:`[force \cdot length]`
-        fx (float): x component of force, retained for backwards compatibility
-          :math:`[\mathrm{force}]`
-        fy (float): y component of force, retained for backwards compatibility
-          :math:`[\mathrm{force}]`
-        fz (float): z component of force, retained for backwards compatibility
-          :math:`[\mathrm{force}]`
-        group (``hoomd.group``): Group for which the force will be set.
-        callback (`callable`): A python callback invoked every time the forces
-            are computed
-
-    :py:class:`constant` specifies that a constant force should be added to
-    every particle in the simulation or optionally to all particles in a group.
+    This class serves as a base class for all custom forces and should not be
+    instantiated directly. A custom force should be derived from this class, and
+    should override a method called `set_forces` in order to set the force data
+    however they wish. Custom forces have direct, zero-copy access to
+    HOOMD-Blue's underlying force buffers via either the
+    `cpu_local_force_arrays` or `gpu_local_force_arrays` property. Choose the
+    property that corresponds to the device you wish to alter the data on. In
+    addition to zero-copy access to force buffers, custom forces have access to
+    the local snapshot API via the ``_state.cpu_local_snapshot`` or the
+    ``_state.gpu_local_snapshot`` property. See the documentation in
+    ``hoomd.state`` for more information on the local snapshot API.
 
-    Note:
-        Forces are kept constant during the simulation. If a callback should
-        re-compute particle forces every time step, it needs to overwrite the
-        old forces of **all** particles with new values.
-
-    Note:
-        Per-particle forces take precedence over a particle group, which takes
-        precedence over constant forces for all particles.
+    When accessing the local force arrays, the force, potential energy, torque,
+    and virial of any particle can be set to any value.
 
     Examples::
 
-        force.constant(fx=1.0, fy=0.5, fz=0.25)
-        const = force.constant(fvec=(0.4,1.0,0.5))
-        const = force.constant(fvec=(0.4,1.0,0.5),group=fluid)
-        const = force.constant(fvec=(0.4,1.0,0.5), tvec=(0,0,1) ,group=fluid)
+        class MyCustomForce(hoomd.force.Custom):
+            def __init__(self):
+                super().__init__()
 
-        def updateForces(timestep):
-            global const
-            const.setForce(tag=1, fvec=(1.0*timestep,2.0*timestep,3.0*timestep))
-        const = force.constant(callback=updateForces)
-    """
+            def set_forces(self, timestep):
+                with self.cpu_local_force_arrays as arrays:
+                    arrays.force = -5
+                    arrays.torque = 3
+                    arrays.potential_energy = 27
+                    arrays.virial = np.arange(6)[None, :]
```

Comment:
```suggestion
                    arrays.force[:] = -5
                    arrays.torque[:] = 3
                    arrays.potential_energy[:] = 27
                    arrays.virial[:] = np.arange(6)[None, :]
```
And similarly for all the other examples.